cosmos/cosmos-sdk
# Migrate API Server

## Description

This PR revises the model in which a API (REST) server is started. The server is now started _in process_ along with the application binary & Tendermint -- everything is in a single process. This drastically reduces various complexities and simplifies the mental model. 

Operators may now enable/disable the API server along with various other options. In the future, they'll be able to provide Caddy/TLS/Proxy configurations as well.

One thing left to consider is if how to handle proof verifications (i.e. trust-node) -- I left a TODO for this.

From an app developer's POV, not much changes. They still provide module route registration, except instead of passing it to a command, they now pass it along with their app constructor to the start command.

To test:

1. `make build-docker-local-simapp` (only needed to be invoked once to build the image)
2. `make clean localnet-start`
3. `BUILDDIR=./build/macos make build-sim`
4. Invoke and test various routes (localhost:1317-1320)
5. `make localnet-stop`

closes: #6408

/cc @jackzampolin @marbar3778 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.
